documentation/ArrayExprLoweringToFIR.md

```diff
@@ -0,0 +1,303 @@
+# Array Expression Lowering
+This document describes how Fortran array expressions will be lowered from their `Fortran::evaluate::Expr` tree representation to a sequence of FIR operations.
+Some higher-level design regarding array expression evaluation is presented in [Array Expression Evaluation Design Document](documentation/ArrayComposition.md).
+The objectives of array expression lowering to FIR are to produce a sequence of FIR operations that:
+1.  Implements Fortran Semantics and f18 semantic extensions
+2.  Exposes enough structure so that array optimizations can later be performed on FIR
+3.  Either implements, or makes it possible to implement as FIR transformations the expressions rewriting exposed in [Array Expression Evaluation Design Document](documentation/ArrayComposition.md).
+## Discussions about the objectives
```

Comment:
I think it would be good so summarize the objectives here before going into detail.  I think that there should be four objectives, two of which are described in this document:
  1) To preserve Fortran semantics
  2) To generate code that is able to take advantage of existing and future LLVM/MLIR optimizations.
  3) To generate code that can be parallelized.  I'm too ignorant of LLVM/MLIR to know if this is the same as the previous objective.
  4) To enable other tools to help programmers debug/optimize/measure/understand the generated code.  At the very least, we'll want to be able to map generated instructions and data back to statements and data declarations and references.

Reply:
I feel like it is slightly repeating what is just above for some points. I agree I said nothing about parallelism and tools though, I will think a bit more about it.